configcat/dart-sdk
# Config v6

### Describe the purpose of your pull request

Implements config v6 support.

Features/improvements:
* [x] Fetch new config json version (`config_v6.json`) and update config model
* [x] SDK key format validation (when client is not set up to use local-only flag overrides)
* [x] AND conditions
* [x] New condition types
   * Prerequisite flag condition
   * Segment condition
* [x] Changed comparison operators
   * (NOT) CONTAINS ANY OF
* [x] New comparison operators
   * BEFORE/AFTER
   * TEXT (NOT) EQUALS (both hashed and cleartext)
   * TEXT (NOT) STARTS/ENDS WITH ANY OF (both hashed and cleartext)
   * ARRAY CONTAINS ANY OF (both hashed and cleartext)
* [x] Percentage options within targeting rules
* [x] Percentage options based on custom attributes
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] User Object attribute/comparison value trimming
* [x] Comparison value hashing changes (SHA256, salting)
* [ ] ~Flag overrides vs. config json salt and segments~ (skipped as JS SDKs don't support full config json flag overrides)
* [x] `EvaluationDetails.matchedTargetingRule`/`matchedPercentageOption` properties (rename + set combinations correctly)
* [x] Update `ConfigCatClient.getKeyAndValueFromSettingsMap`

Tests: 
* [x] Update cache key/payload tests
* [x] Evaluation matrix tests
  * Config v1 features (both V1 & V2)
  * Config v2 features
* [x] Evaluation log tests
* [x] SDK key format validation
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] Flag overrides vs. prerequisite flags (circular dependency, comparison value type mismatch, dependent/prereq flag overriding)
* [ ] ~Flag overrides vs. config json salt and segments~ (skipped as JS SDKs don't support full config json flag overrides)
* [x] `EvaluationDetails.matchedTargetingRule`/`matchedPercentageOption` properties (rename + set combinations correctly)
* [x] Special character tests (https://trello.com/c/qh4jHrvm)
* [x] Comparison attribute and comparison value trimming tests (https://trello.com/c/vt8FFrSZ)
* [x] Comparison attribute conversion to canonical string representation tests (https://trello.com/c/CZgGrrcb)

Related improvements/fixes:
* [ ] ~Dynamically typed languages should log a warning when default value doesn't match the setting type (see https://trello.com/c/Le6vimGu)~
* [ ] We should go to the cache in all polling modes instead of using the in memory variable (see https://trello.com/c/rreKm64A)
* [x] Improve message of log event 1103 (https://trello.com/c/Mb7rvm64)

Breaking changes
- Change the `ConfigCatUser` constructor `custom` parameter type to Map<String, Object> to allow other values than string.
- Rename the `matchedEvaluationRule` property to `matchedTargetingRule` and the `matchedEvaluationPercentageRule` property to `matchedPercentageOption` in `EvaluationDetails`.
- Change `Config` model

### Requirement checklist (only if applicable)
